bugfix: pointer check failure in fgets

mauguignard opened the pull request:
Now we assume str_length greater than 0, since no dereference should take place when str_length equals 0.
Updated cbmc-library/fgets-01 regression test.

Fixes: #4621

- [x] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

mauguignard committed:
bugfix: pointer check failure in fgets

Now we assume str_length greater than 0, since no dereference should take place when str_length equals 0.
Updated cbmc-library/fgets-01 regression test.

Fixes: #4621

kroening commented:
May I suggest to switch this to using `__CPROVER_w_ok(s, n)`?
Say something like
`__CPROVER_precondition(__CPROVER_w_ok(str, str_length+1),
                         "fgets buffer writeable");`

mauguignard committed:
remove explicit pointer dereference in fgets

Now, a w_ok precondition is used to check if the buffer is writable.

mauguignard committed:
simplified buffer precondition in fgets

Since we are interested in checking if the buffer is writable and we are
assuming that str_length < size, verify __CPROVER_w_ok(str, str_length + 1)
is the same as verify __CPROVER_w_ok(str, size).
